E2E Sample - Live Stream Video Object Detection with Onnx in WPF Desktop App (#602)
dotnet/machinelearning-samples: samples/csharp/end-to-end-apps/ObjectDetection-Onnx/README.md

![Animated image showing object detection web sample](./docs/Screenshots/ObjectDetection.gif)

Alternatively you can try uploading your own images as shown below.

![Animated image showing object detection web sample](./docs/Screenshots/FileUpload.gif)

## Pre-trained model

There are multiple pre-trained models for identifying multiple objects in the images. Both the **WPF app** and the **Web app** use the pre-trained model, **Tiny Yolo2** in [**ONNX**](http://onnx.ai/) format. This model is a real-time neural network for object detection that detects 20 different classes. It is made up of 9 convolutional layers and 6 max-pooling layers and is a smaller version of the more complex full [YOLOv2](https://pjreddie.com/darknet/yolov2/) network.

The Open Neural Network eXchange i.e [ONNX](http://onnx.ai/) is an open format to represent deep learning models. With ONNX, developers can move models between state-of-the-art tools and choose the combination that is best for them. ONNX is developed and supported by a community of partners, including Microsoft.

> The model is downloaded from the [ONNX Model Zoo](https://github.com/onnx/models/tree/master/tiny_yolov2) which is a is a collection of pre-trained, state-of-the-art models in the ONNX format.

Author: Resolved in 368f84c2f15116ae98ddd8d797e326e953544dcc

public class ImageInputData
{
    [ImageType(416, 416)]
    public Bitmap Image { get; set; }
}
```

### ML.NET: Configure the model

The first step is to create an empty DataView as we just need schema of data while configuring up model.

```csharp
var dataView = _mlContext.Data.LoadFromEnumerable(new List<ImageInputData>());
```

The second step is to define the estimator pipeline. Usually when dealing with deep neural networks, you must adapt the images to the format expected by the network. For this reason, the code below resizes and transforms the images (pixel values are normalized across all R,G,B channels).

```csharp
var pipeline = _mlContext.Transforms.ResizeImages(resizing: ImageResizingEstimator.ResizingKind.Fill, outputColumnName: "image", imageWidth: ImageSettings.imageWidth, imageHeight: ImageSettings.imageHeight, inputColumnName: nameof(ImageInputData.Image))
                .Append(_mlContext.Transforms.ExtractPixels(outputColumnName: "image"))
                .Append(_mlContext.Transforms.ApplyOnnxModel(modelFile: onnxModelFilePath, outputColumnNames: new[] { TinyYoloModelSettings.ModelOutput }, inputColumnNames: new[] { TinyYoloModelSettings.ModelInput }));
```

> You also need to inspect the neural network to get the names of the input and output nodes. To do this, you can use tools like [Netron](https://github.com/lutzroeder/netron), which is automatically installed with [Visual Studio Tools for AI](https://visualstudio.microsoft.com/downloads/ai-tools-vs/).

Reviewer: Is Model Builder (which is where this link leads) different than Visual Studio Tools for AI? If we're just referring to Model Builder, maintain consistency and just call it Model Builder.

```suggestion
You also need to inspect the neural network to get the names of the input and output nodes. To do this, you can use tools like [Netron](https://github.com/lutzroeder/netron), which is automatically installed with [Model Builder](https://visualstudio.microsoft.com/downloads/ai-tools-vs/), a VS extension to make ML more approachable.
```

> These names are used later when we define the estimation pipeline: in the case of the Tiny Yolo2 network, the input tensor is named **'image'** and the output is named **'grid'**.

Reviewer: ```suggestion
These names are used later when we define the estimation pipeline: in the case of the Tiny Yolo2 network, the input tensor is named **'image'** and the output is named **'grid.'**
```

We'll use these to define the **input** and **output** parameters of the Tiny Yolo2 Onnx Model.

```csharp
public struct TinyYoloModelSettings
{

>     // for checking TIny yolo2 Model input and output parameter names,

Reviewer: ```suggestion
    // To check Tiny Yolo2 Model input and output parameter names,
```

>     // you can use tools like Netron, which is installed by Visual Studio AI Tools

Reviewer: Again, depends on whether we mean model builder or a different/broader toolset

Author: Resolved in 368f84c2f15116ae98ddd8d797e326e953544dcc

>     // input tensor name

Reviewer: ```suggestion
    // Input tensor name
```

    public const string ModelInput = "image";

>     // output tensor name

Reviewer: ```suggestion
    // Output tensor name
```

    public const string ModelOutput = "grid";
}
```

> 

Reviewer: I would add a sentence before or after this image to explain/reinforce what it is showing.

Author: Resolved 368f84c2f15116ae98ddd8d797e326e953544dcc

![inspecting neural network with netron](./docs/Netron/netron.PNG)

Create the model by fitting the DataView.

```csharp
var model = pipeline.Fit(dataView);
```

## Load model and create PredictionEngine

> After the model is configured, we need to save the model, load the saved model, create a `PredictionEngine`, and the pass the image to the engine to detect objects using the model.

Reviewer: ```suggestion
After the model is configured, we need to save the model, load the saved model, create a `PredictionEngine`, and then pass the image to the engine to detect objects using the model.
```

> This is one place that the **Web** app and the **WPF** app differ slightly.  The **Web** app uses a `PredicitonEnginePool` to efficiently manage and provide the service with a `PredictionEngine` to use to make predictions.

Reviewer: Separate out the two sentences so that users can visually see a separation between the web app vs desktop app explanations.

```suggestion
This is one place that the **Web** app and the **WPF** app differ slightly.  

The **Web** app uses a `PredicitonEnginePool` to efficiently manage and provide the service with a `PredictionEngine` to use to make predictions.
```

```csharp
public ObjectDetectionService(PredictionEnginePool<ImageInputData, ImageObjectPrediction> predictionEngine)
{
    this.predictionEngine = predictionEngine;
}
```

Whereas the **WPF** desktop app creates a single `PredictionEngine` and caches locally to be used for each frame prediction.

```csharp
public PredictionEngine<ImageInputData, ImageObjectPrediction> GetMlNetPredictionEngine()
{
    return _mlContext.Model.CreatePredictionEngine<ImageInputData, ImageObjectPrediction>(_mlModel);
}
```

## Detect objects in the image

> When obtaining the prediction, we get an `float` array of size **21125** in the `PredictedLabels` property. This is the 125x13x13 output of the model discussed earlier. We then use the `YoloOutputParser` class to interpret and returns a number of bounding boxes for each image. Again these boxes are filtered so that we retrieve only 5 with high confidence.

Reviewer: ```suggestion
When obtaining the prediction, we get an `float` array of size **21125** in the `PredictedLabels` property. This is the 125x13x13 output of the model discussed earlier. We then use the `YoloOutputParser` class to interpret and return a number of bounding boxes for each image. Again, these boxes are filtered so that we retrieve only 5 with high confidence.
```
